spectral_cube/spectral_cube.py

```diff
@@ -1912,6 +1912,122 @@ def subcube_from_ds9region(self, ds9region, allow_empty=False):
         # that we can now crop out
         return masked_subcube.minimal_subcube(spatial_only=True)
 
+    def subcube_from_ds9region_new(self, ds9region, allow_empty=False):
+        """
+        Extract a masked subcube from a ds9 region
+        (only functions on celestial dimensions)
+
+        Parameters
+        ----------
+        ds9region: str
+            The region to extract
+        allow_empty: bool
+            If this is False, an exception will be raised if the region
+            contains no overlap with the cube
+        """
+        import regions
+
+        if isinstance(ds9region, six.string_types):
+            ds9region = regions.DS9Parser(ds9region).shapes.to_regions()
+        else:
+            raise TypeError("{0} should be a DS9 string".format(ds9region))
+
+        return self.subcube_from_regions(ds9region, allow_empty)
+
+    def subcube_from_crtf(self, region, allow_empty=False):
+        """
+        Extract a masked subcube from a CRTF region
+        (only functions on celestial dimensions)
+
+        Parameters
+        ----------
+        region: str
+            The CRTF region string to extract
+        allow_empty: bool
+            If this is False, an exception will be raised if the region
+            contains no overlap with the cube
+        """
+        import regions
+
+        if isinstance(region, six.string_types):
+            region = regions.CRTFParser(region).shapes.to_regions()
+        else:
+            raise TypeError("{0} should be a CRTF string".format(region))
+
+        return self.subcube_from_regions(region, allow_empty)
+
+    def subcube_from_regions(self, region_list, allow_empty=False):
+        """
+        Extract a masked subcube from a list of `regions.Region` object
+        (only functions on celestial dimensions)
+
+        Parameters
+        ----------
+        region_list: `regions.Region` list
+            The region to extract
+        allow_empty: bool
+            If this is False, an exception will be raised if the region
+            contains no overlap with the cube
+        """
+        import regions
+
+        # Every region should be a `regions.PixelRegion` object.
+        regs = []
+        for x in region_list:
+            if isinstance(x, regions.SkyRegion):
+                regs.append(x.to_pixel(self.wcs.celestial))
+            elif isinstance(x, regions.PixelRegion):
+                regs.append(x)
+            else:
+                raise TypeError("'{}' should be `regions.Region` object".format(x))
+
+        # List of regions are converted to a `regions.CompoundPixelRegion` object.
+        compound_region = _compound_region(regs)
+
+        # Compound mask of all the regions.
+        mask = compound_region.to_mask()
+
+        # Checking for frequency/velocity range of each region.
+        ranges = [x.meta.get('range', None) for x in regs]
```

Comment:
ooh, this is a bit ugly, since they could technically be different....  we will need additional machinery to create the masks for each PPV sub-region instead of using the compound region (or at least, instead of *just* using the compound region)

Reply:
Oh! Can you be a little more elaborative here? Like what technical differences can we face here and what machinery should we use to create the masks? Currently it just replicates the old `subcube_from_ds9region`.

Reply:
Right, but the old `subcube_from_ds9region` only did spatial.  With CRTF, we can do spatial-spatial-spectral.  So this is fine for the ds9 regions, but we want the CRTFs to allow us to do more.

Reply:
Do the spectral dimensions have any system or convention ? I just try to find out the spectral slab range from the `range` key. What can we do more? You were also talking about technical differences?

Reply:
Also, we can make use of the stokes parameters in CRTF regions.

Reply:
Yes, there is some important frame information in the `range`.  The `velocity_convention` is something that should be specified in the CRTF region... I'm not sure what the default is.  

So, for example, if you have a region with `range=[10 km /s, 20 km / s]` but the cube is in frequency (GHz, say), you would need to do `range.to(u.GHz, u.doppler_radio(rest_frequency)` if the velocity convention is radio...

Reply:
`veltype` is for the velocity convention. The dafault value is the image value. 
`restfreq` is for rest frequency. The default says image value .
I will work on this.

Reply:
https://help.almascience.org/index.php?/Knowledgebase/Article/View/86/0/what-are-the-frequency-reference-frames-in-casa I got something here.

Reply:
Now, I get why we need a machinery here. Let's just see if astropy has handle all this.